### PR TITLE
demo: service worker for offline access

### PR DESCRIPTION
**Pull request by divdavem**

### CODE REVIEW

File: demo/src/service-worker.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+			await self.skipWaiting();
+		})()
+	);
+});
+
+self.addEventListener('activate', (event) => {
+	event.waitUntil(
+		(async () => {
+			await self.clients.claim();
+			for (const key of await caches.keys()) {
```
Comment (quentinderoubaix) [inline demo/src/service-worker.ts:47]: If ```caches``` allows it, you could delete all deprecated caches in parallel:
```typescript
await Promise.all((await caches.keys()).filter((key) => key !== CACHE).map((key) => caches.delete(key)));
```

File: demo/scripts/copySamples.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+			for (const framework of ['angular', 'react', 'svelte']) {
+				const src = path.join(__dirname, `../../${framework}/dist/demo/`);
+				const dst = `${framework}/samples/`;
+				const files = await glob('**/*', {
+					cwd: src,
+					nodir: true,
+				});
+				for (const file of files) {
+					const name = path.posix.join(dst, file);
+					console.log(name);
```
Comment (quentinderoubaix) [inline demo/scripts/copySamples.plugin.ts:20]: You could use instead a logger created using vite ```createLogger``` utility.

File: demo/src/service-worker.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	return {withHash, withoutHash};
+};
+
+self.addEventListener('install', (event) => {
+	event.waitUntil(
+		(async () => {
+			const cache = await caches.open(CACHE);
+			const {withHash, withoutHash} = splitAssets(ASSETS);
+			const missingWithHash: string[] = [];
+			for (const url of withHash) {
```
Comment (quentinderoubaix) [inline demo/src/service-worker.ts:29]: Maybe could be written as the following:
```typescript
await Promise.all(
	withHash.map((url) =>
		caches.match(url).then((response) => {
			if (response?.ok) {
				cache.put(url, response);
			} else {
				missingWithHash.push(url);
			}
		})
	)
);
```
Not sure of the gain however.

**Comment by divdavem** [comment]
@quentinderoubaix Thank you for your good suggestions. I have added a commit to implement these changes.

**Review by quentinderoubaix — APPROVED** [review]
LGTM !

**Pull request merged**